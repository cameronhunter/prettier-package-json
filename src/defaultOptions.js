module.exports = Object.freeze({
  useTabs: false,
  tabWidth: 2,
  expandUsers: false,
  keyOrder: [
    /**
     * Details
     */
    '$schema',
    'private',
    'name',
    'description',
    'license',
    'author',
    'maintainers',
    'contributors',
    'homepage',
    'repository',
    'bugs',
    'version',
    'type',

    /**
     * Yarn specific
     */
    'workspaces',

    /**
     * Configuration
     */
    'main',
    'module',
    'browser',
    'man',
    'preferGlobal',
    'bin',
    'files',
    'directories',
    'scripts',
    'config',
    'sideEffects',
    'types',
    'typings',

    /**
     * Dependencies
     */
    'optionalDependencies',
    'dependencies',
    'bundleDependencies',
    'bundledDependencies',
    'peerDependencies',
    'devDependencies',

    /**
     * Used for npm search
     */
    'keywords',

    /**
     * Constraints
     */
    'engines',
    'engine-strict',
    'engineStrict',
    'os',
    'cpu',

    /**
     * Package publishing configuration
     */
    'publishConfig'
  ]
});
