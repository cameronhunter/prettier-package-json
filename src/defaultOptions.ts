import { Options } from './types';

export const defaultOptions: Options = {
  useTabs: false,
  tabWidth: 2,
  expandUsers: false,
  keyOrder: [
    /**
     * Details
     */
    '$schema',
    'private',
    'name',
    'version',
    'description',
    'license',
    'author',
    'maintainers',
    'contributors',
    'homepage',
    'repository',
    'bugs',
    'type',

    /**
     * Configuration
     */
    'exports',
    'main',
    'module',
    'browser',    
    'man',
    'preferGlobal',
    'bin',
    'files',
    'directories',
    'scripts',
    'config',
    'sideEffects',
    'types',
    'typings',

    /**
     * Yarn specific
     */
    'workspaces',
    'resolutions',

    /**
     * Dependencies
     */
    'dependencies',
    'bundleDependencies',
    'bundledDependencies',
    'peerDependencies',
    'peerDependenciesMeta',
    'optionalDependencies',
    'devDependencies',

    /**
     * Used for npm search
     */
    'keywords',

    /**
     * Constraints
     */
    'engines',
    'engineStrict',
    'os',
    'cpu',

    /**
     * Package publishing configuration
     */
    'publishConfig'
  ]
};
