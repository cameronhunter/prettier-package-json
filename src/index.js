
const defaults = require('./defaultOptions');
const sortObject = require('sort-object-keys');
const sortScripts = require('./sort-scripts');
const sortUsers = require('./sort-contributors');
const sortFiles = require('./sort-files');

function stringify(object, options) {
  const space = options.useTabs ? '\t' : options.tabWidth;
  return JSON.stringify(object, null, space) + '\n';
}

function format(packageJson, opts) {
  const options = Object.assign({}, defaults, opts);

  const json = Object.assign(
    {},
    packageJson,
    sortUsers('author', packageJson, opts),
    sortUsers('contributors', packageJson, opts),
    sortUsers('maintainers', packageJson, opts),
    sort('man', packageJson),
    sort('bin', packageJson),
    sortFiles(packageJson),
    sort('directories', packageJson),
    sortScripts(packageJson.scripts),
    sort('config', packageJson),
    sort('optionalDependencies', packageJson),
    sort('dependencies', packageJson),
    sort('bundleDependencies', packageJson),
    sort('bundledDependencies', packageJson),
    sort('peerDependencies', packageJson),
    sort('devDependencies', packageJson),
    sort('keywords', packageJson),
    sort('engines', packageJson),
    sort('os', packageJson),
    sort('cpu', packageJson),
    sort('publishConfig', packageJson)
  );

  return stringify(sortObject(json, options.keyOrder), options);
}

function check(packageJson, opts) {
  try {
    const options = Object.assign({}, defaults, opts);
    const isString = typeof packageJson === 'string';
    const object = isString ? JSON.parse(packageJson) : packageJson;
    const original = isString ? packageJson : stringify(object, options);
    const formatted = format(object, options);
    return stringify(object, options) === formatted;
  } catch(e) {
    return false;
  }
}

function sort(key, packageJson) {
  const input = packageJson[key] || {};
  if (Array.isArray(input)) {
    if (input.length === 0) {
      return {};
    } else {
      return { [key]: input.sort() };
    }
  } else if (typeof input === 'object') {
    const keys = Object.keys(input);
    if (keys.length === 0) {
      return {};
    } else {
      return { [key]: sortObject(input, keys.sort()) };
    }
  } else {
    return {};
  }
}

module.exports = {
  format,
  check
};
