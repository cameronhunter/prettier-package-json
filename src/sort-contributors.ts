import parse from 'parse-author';
import sortObject from 'sort-object-keys';
import orderBy from 'sort-order';
import { Author, Options, PackageJson } from './types';

// prettier-ignore
// Sort by a field in an object
const field = <TType extends object>(name: keyof TType) => (...args: [TType, TType]) => {
  const [a, b] = args.map((obj) => obj[name]);
  if (a !== b) {
    return a < b ? -1 : 1;
  } else {
    return 0;
  }
};

export default function sortContributors<TKey extends 'author' | 'contributors' | 'maintainers'>(
  key: TKey,
  packageJson: PackageJson,
  opts: Options = {}
) {
  const sortByName = field<Author>('name');
  const sortByEmail = field<Author>('email');

  const contributors = hydrate(packageJson[key])
    .map((user) => sortObject(user, ['name', 'email', 'url']))
    .sort(orderBy(sortByName, sortByEmail));

  switch (contributors.length) {
    case 0:
      return {};
    case 1:
      if (!opts.enforceMultiple) {
        return { [key]: opts.expandUsers ? contributors[0] : stringify(contributors[0]) };
      }
    default:
      return { [key]: opts.expandUsers ? contributors : contributors.map(stringify) };
  }
}

function hydrate(input: undefined | string | Author | Array<string | Author>): Author[] {
  if (!input) {
    return [];
  } else if (Array.isArray(input)) {
    return input.reduce((state, item) => [...state, ...hydrate(item)], [] as Author[]);
  } else if (typeof input === 'object') {
    return [input];
  } else if (typeof input === 'string') {
    return [parse(input)];
  } else {
    return [];
  }
}

function stringify(user: Author): string {
  const name = user.name;
  const email = user.email ? `<${user.email}>` : undefined;
  const url = user.url ? `(${user.url})` : undefined;

  return [name, email, url].filter(Boolean).join(' ');
}
