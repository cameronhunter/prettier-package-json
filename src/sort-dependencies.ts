import sortObject from 'sort-object-keys';
import { PackageJson } from './types';

function order(a: string, b: string) {
  return a.localeCompare(b, 'en');
}

export default function sortDependencies<
  TKey extends
    | 'bundledDependencies'
    | 'bundleDependencies'
    | 'dependencies'
    | 'devDependencies'
    | 'optionalDependencies'
    | 'peerDependencies'
>(key: TKey, packageJson: PackageJson) {
  const dependencies = packageJson[key];
  const keys = Object.keys(dependencies || {}) as Array<keyof typeof dependencies & string>;
  return keys.length === 0 ? {} : { [key]: sortObject(dependencies, keys.sort(order)) };
}
