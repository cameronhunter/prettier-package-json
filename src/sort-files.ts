/**
 * Sort and filter files field
 *
 * More info:
 *   https://docs.npmjs.com/files/package.json#files
 */

import minimatch from 'minimatch';
import { PackageJson } from './types';

type FilterFn<T> = (...args: T[]) => boolean;

const not = <T>(filterFn: FilterFn<T>) => (arg: T) => !filterFn(arg);
const or = <T>(...filterFns: FilterFn<T>[]) => (arg: T) => filterFns.some((fn) => fn(arg));

const ALWAYS_INCLUDED = [
  /^package.json$/,
  /^README.*/i,
  /^CHANGE(S|LOG).*/i,
  /^HISTORY.*/i,
  /^LICEN(C|S)E.*/i,
  /^NOTICE.*/i
]
  .map((regex) => (filepath: string) => regex.test(filepath))
  .reduce((a, b) => or(a, b));

const ALWAYS_EXCLUDED = [
  '.git',
  'CVS',
  '.svn',
  '.hg',
  '.lock-wscript',
  '.wafpickle-N',
  '.*.swp',
  '.DS_Store',
  '._*',
  'npm-debug.log',
  '.npmrc',
  'node_modules',
  'config.gypi',
  '*.orig',
  'package-lock.json'
]
  .map((glob) => (minimatch.filter(glob) as any) as FilterFn<string>)
  .reduce((a, b) => or(a, b));

export default function sortFiles(packageJson: PackageJson): { files?: PackageJson['files'] } {
  const { files = [], main } = packageJson;

  const isPackageMain = (filepath: string) => filepath === main;
  const ignored = or(ALWAYS_INCLUDED, ALWAYS_EXCLUDED, isPackageMain);
  const directoriesFirst = (a: string, b: string) => {
    if (a.endsWith('/') && !b.endsWith('/')) {
      return -1;
    } else if (!a.endsWith('/') && b.endsWith('/')) {
      return 1;
    } else {
      return 0;
    }
  };
  const exclusionsLast = (a: string, b: string) => {
    if (a.startsWith('!') && !b.startsWith('!')) {
      return 1;
    } else if (!a.startsWith('!') && b.startsWith('!')) {
      return -1;
    } else {
      return 0;
    }
  };

  const sortedAndFilteredFiles = files.filter(not(ignored)).sort().sort(directoriesFirst).sort(exclusionsLast);

  return sortedAndFilteredFiles.length > 0 ? { files: sortedAndFilteredFiles } : {};
}
