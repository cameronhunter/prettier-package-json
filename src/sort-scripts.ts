import sortObject from 'sort-object-keys';
import orderBy from 'sort-order';
import { PackageJson } from './types';

const PRE_OR_POST_PREFIX = /^(pre|post)/;
const NPM_RUN_ALL_SEPARATOR = /([:/])/;
const NPM_BUILTIN_SCRIPTS = ['install', 'prepare', 'pack', 'publish', 'restart', 'start', 'stop', 'test', 'version', 'uninstall'];

interface ParsedScriptName {
  prefix?: 'pre' | 'post';

  /**
   * Base event name in the npm lifecycle. Often the string with prefix removed.
   * For example, for an input `preprettier:lint`, the base would be `prettier:lint`.
   */
  base: string;


  /**
   * The full original name of the script
   */
  original: string;
};

function parseScriptName(scripts: PackageJson['scripts'], original: string): ParsedScriptName {
  // prepublishOnly is a script whose base event is publish.
  // cf. https://docs.npmjs.com/cli/v9/using-npm/scripts#npm-publish
  if (original === 'prepublishOnly') {
    return { prefix: 'pre', base: 'publish', original };
  }

  const prefixMatch = PRE_OR_POST_PREFIX.exec(original);
  const prefix = prefixMatch ? prefixMatch[0] as 'pre' | 'post' : undefined;
  const base = prefix ? original.slice(prefix.length) : original;

  if (scripts?.hasOwnProperty(base) || NPM_BUILTIN_SCRIPTS.includes(base)) {
    return { prefix, base, original };
  }

  return { prefix: undefined, base: original, original };
}

// Sort alphabetically by script name excluding pre/post prefixes
function scriptName(scripts: PackageJson['scripts']) {
  return function (a: string, b: string): 1 | 0 | -1 {
    const aParsed = parseScriptName(scripts, a);
    const bParsed = parseScriptName(scripts, b);

    // For example, for an input name `prettier:lint`, the segments would be `['prettier', ':', 'lint']`.
    // The segments contains separators. There are two types of separators, to determine their order.
    const aSegments = aParsed.base.split(NPM_RUN_ALL_SEPARATOR);
    const bSegments = bParsed.base.split(NPM_RUN_ALL_SEPARATOR);

    const aLength = aSegments.length;
    const bLength = bSegments.length;

    // Compare each segment, and when the strings are different, return 1 or -1 in alphabetical order.
    for (let i = 0; i < Math.min(aLength, bLength); i++) {
      if (aSegments[i] !== bSegments[i]) {
        return aSegments[i] < bSegments[i] ? -1 : 1;
      }
    }

    // Compare segments length, return 1 or -1 in ascending order of length.
    if (aLength !== bLength) {
      return aLength < bLength ? -1 : 1;
    }
    return 0;
  }
}

// Sort by pre, script, post
function prePostHooks(scripts: PackageJson['scripts']) {
  return function (a: string, b: string): 1 | 0 | -1 {
    const aParsed = parseScriptName(scripts, a);
    const bParsed = parseScriptName(scripts, b);
    if (aParsed.prefix === 'pre' || bParsed.prefix === 'post') {
      return -1;
    } else if (aParsed.prefix === 'post' || bParsed.prefix === 'pre') {
      return 1;
    } else if (aParsed.base !== bParsed.base) {
      return aParsed.base < bParsed.base ? -1 : 1;
    }
    return 0;
  }
}

export default function sortScripts(scripts: PackageJson['scripts']): { scripts?: PackageJson['scripts'] } {
  const keys = Object.keys(scripts || {}) as Array<keyof PackageJson['scripts']>;
  const order = orderBy(scriptName(scripts), prePostHooks(scripts));
  return keys.length === 0 ? {} : { scripts: sortObject(scripts, keys.sort(order)) };
}
