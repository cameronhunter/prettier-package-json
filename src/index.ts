import { cosmiconfigSync } from 'cosmiconfig';
import sortObject from 'sort-object-keys';
import { defaultOptions } from './defaultOptions';
import sortDependencies from './sort-dependencies';
import sortUsers from './sort-contributors';
import sortFiles from './sort-files';
import sortScripts from './sort-scripts';
import type { Options, PackageJson, PackageJsonKey } from './types';

export { Options };

export function format(packageJson: PackageJson, opts?: Options) {
  const options = getConfig(opts);

  const json = {
    ...packageJson,
    ...sortUsers('author', packageJson, opts),
    ...sortUsers('maintainers', packageJson, { ...opts, enforceMultiple: true }),
    ...sortUsers('contributors', packageJson, { ...opts, enforceMultiple: true }),
    ...sort('man', packageJson),
    ...sort('bin', packageJson),
    ...sortFiles(packageJson),
    ...sort('directories', packageJson),
    ...sortScripts(packageJson.scripts),
    ...sort('config', packageJson),
    ...sortDependencies('optionalDependencies', packageJson),
    ...sortDependencies('dependencies', packageJson),
    ...sortDependencies('bundleDependencies', packageJson),
    ...sortDependencies('bundledDependencies', packageJson),
    ...sortDependencies('peerDependencies', packageJson),
    ...sortDependencies('devDependencies', packageJson),
    ...sort('keywords', packageJson),
    ...sort('engines', packageJson),
    ...sort('os', packageJson),
    ...sort('cpu', packageJson),
    ...sort('publishConfig', packageJson)
  };

  return stringify(sortObject(json, options.keyOrder), options);
}

export function check(packageJson: string | PackageJson, opts: Options): boolean {
  try {
    const options = getConfig(opts);
    const object = typeof packageJson === 'string' ? JSON.parse(packageJson) : packageJson;
    const formatted = format(object, options);
    return stringify(object, options) === formatted;
  } catch (e) {
    return false;
  }
}

function getConfig(options?: Options): Options {
  const explorer = cosmiconfigSync('prettier-package-json');
  const result = explorer.search();
  return { ...defaultOptions, ...result?.config, ...options };
}

function stringify(object: object, options: Options): string {
  const space = options.useTabs ? '\t' : options.tabWidth;
  return JSON.stringify(object, null, space) + '\n';
}

function sort(key: PackageJsonKey, packageJson: PackageJson): PackageJson {
  const input = packageJson[key];

  if (Array.isArray(input)) {
    return input.length === 0 ? {} : { [key]: input.sort() };
  }

  if (typeof input === 'object') {
    const keys = Object.keys(input) as Array<keyof typeof input>;
    return keys.length === 0 ? {} : { [key]: sortObject(input, keys.sort()) };
  }

  return {};
}
