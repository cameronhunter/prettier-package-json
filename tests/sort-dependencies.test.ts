import { format } from '../src';

test('It orders dependencies in alphabetical order', () => {
  const json = {
    dependencies: {
        foo_bar: "3",
        g: "4",
        a: "1",
        "foo-bar": "2", // the "-" should be before the "_" to match `npm` sorting
    }
  };

  expect(format(json)).toMatchSnapshot();
});
