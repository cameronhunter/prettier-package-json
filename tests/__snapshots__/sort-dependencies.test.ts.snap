// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It orders dependencies in alphabetical order 1`] = `
"{
  \\"dependencies\\": {
    \\"a\\": \\"1\\",
    \\"foo-bar\\": \\"2\\",
    \\"foo_bar\\": \\"3\\",
    \\"g\\": \\"4\\"
  }
}
"
`;
